daemon/pipeline-module.c

```diff
@@ -0,0 +1,429 @@
+        GDBusMethodInvocation *invoc,
+        gint64 id, gpointer user_data)
+{
+  int result = 0;
+  GstStateChangeReturn sc_ret;
+  pipeline_s *p = NULL;
+
+  G_LOCK (pipeline_table_lock);
+  p = (pipeline_s *) g_hash_table_lookup (pipeline_table, GINT_TO_POINTER (id));
+
+  if (!p) {
+    _E ("there is no pipeline with id %" G_GINT64_FORMAT, id);
+    G_UNLOCK (pipeline_table_lock);
+    result = -EINVAL;
+  } else {
+    g_mutex_lock (&p->lock);
+    G_UNLOCK (pipeline_table_lock);
+    sc_ret = gst_element_set_state (p->element, GST_STATE_PLAYING);
+    g_mutex_unlock (&p->lock);
+
+    if (sc_ret == GST_STATE_CHANGE_FAILURE) {
+      _E ("Failed to set the state of the pipline to PLAYING whose id is %" G_GINT64_FORMAT, id);
+      result = -ESTRPIPE;
+    }
+  }
+
+  machinelearning_service_pipeline_complete_start_pipeline (obj, invoc, result);
+
+  return TRUE;
+}
+
+/**
+ * @brief Stop the pipeline with given id. Return the call result.
+ */
+static gboolean dbus_cb_core_stop_pipeline (MachinelearningServicePipeline *obj,
+        GDBusMethodInvocation *invoc,
+        gint64 id, gpointer user_data)
+{
+  int result = 0;
+  GstStateChangeReturn sc_ret;
+  pipeline_s *p = NULL;
+
+  G_LOCK (pipeline_table_lock);
+  p = (pipeline_s *) g_hash_table_lookup (pipeline_table, GINT_TO_POINTER (id));
+
+  if (!p) {
+    _E ("there is no pipeline with id %" G_GINT64_FORMAT, id);
+    G_UNLOCK (pipeline_table_lock);
+    result = -EINVAL;
+  } else {
+    g_mutex_lock (&p->lock);
+    G_UNLOCK (pipeline_table_lock);
+    sc_ret = gst_element_set_state (p->element, GST_STATE_PAUSED);
+    g_mutex_unlock (&p->lock);
+
+    if (sc_ret == GST_STATE_CHANGE_FAILURE) {
+      _E ("Failed to set the state of the pipline to PAUSED whose id is %" G_GINT64_FORMAT, id);
+      result = -ESTRPIPE;
+    }
+  }
+
+  machinelearning_service_pipeline_complete_stop_pipeline (obj, invoc, result);
+
+  return TRUE;
+}
+
+/**
+ * @brief Destroy the pipeline with given id. Return the call result.
+ */
+static gboolean dbus_cb_core_destroy_pipeline (MachinelearningServicePipeline *obj,
+        GDBusMethodInvocation *invoc,
+        gint64 id, gpointer user_data)
+{
+  int result = 0;
+  GstStateChangeReturn sc_ret;
+  pipeline_s *p = NULL;
+
+  G_LOCK (pipeline_table_lock);
+  p = (pipeline_s *) g_hash_table_lookup (pipeline_table, GINT_TO_POINTER (id));
+
+  if (!p) {
+    _E ("there is no pipeline with id %" G_GINT64_FORMAT, id);
+    result = -EINVAL;
+    G_UNLOCK (pipeline_table_lock);
+  } else {
+    g_mutex_lock (&p->lock);
+    G_UNLOCK (pipeline_table_lock);
+    sc_ret = gst_element_set_state (p->element, GST_STATE_NULL);
+    g_mutex_unlock (&p->lock);
+
+    if (sc_ret == GST_STATE_CHANGE_FAILURE) {
+      _E ("Failed to set the state of the pipeline to NULL whose id is %" G_GINT64_FORMAT, id);
+      result = -ESTRPIPE;
+    }
+
+    G_LOCK (pipeline_table_lock);
+    g_hash_table_remove (pipeline_table, GINT_TO_POINTER (id));
```

Comment:
Don't you need to lock p->lock while you remove it from the table?
Besides, because this is a sequence of read and update for table, pipeline_table_lock should be kept from LINE 212 to LINE 232. Calling "remove" on the same id from another thread can break the context.

Reply:
@myungjoo Thanks! I changed the code to keep the `pipeline_table_lock` during the whole sequence.
The `p->lock` is checked in the destroy callback `_pipeline_free`.